### PR TITLE
SwitchAnalyzer - fix for semicolon after case/default

### PR DESCRIPTION
**Pull request by kubawerlos**
See the test cases "two case and default with semicolon instead of colon" and "alternative syntax nested with mixed colon/semicolon".

### CODE REVIEW

File: src/Tokenizer/Analyzer/SwitchAnalyzer.php [hunk 1]
```diff
@@ -94,30 +82,59 @@ private function getCasesEnd(Tokens $tokens, $casesStartIndex)
...
-                continue;
-            }
-            $index = $this->getCasesStart($tokens, $index);
-            if ($tokens[$index]->equals(':')) {
-                ++$depth;
+        $afterEndswitchIndex = $tokens->getNextMeaningfulToken($index);
+
+        $afterEndswitchToken = $tokens[$afterEndswitchIndex];
+
+        return $afterEndswitchToken->equals(';') ? $afterEndswitchIndex : $index;
```
Comment (SpacePossum) [inline src/Tokenizer/Analyzer/SwitchAnalyzer.php:98]: I think `->equalsAny([';', [T_CLOSE_TAG]])`
to support

```php
<?php switch ($foo) : case 10;
                switch ($bar) : case "a": return "b"; case "c"; return "d"; case "e": return "f"; endswitch;
                return;
                case 100: return false; endswitch ?>  <?php echo 1;
```

Comment (kubawerlos) [reply src/Tokenizer/Analyzer/SwitchAnalyzer.php:98]: This is already supported - if after switch token we have closing tag then the end of switch is the switch token, not the closing tag. I've added a test case to show it, see it, and the one above to see the difference in switch end.

**Review by GrahamCampbell — APPROVED** [review]

File: src/Tokenizer/Analyzer/SwitchAnalyzer.php (hunk 1 above)
Comment (SpacePossum) [reply src/Tokenizer/Analyzer/SwitchAnalyzer.php:98]: Thanks for the test!
Maybe I just don't understand the meaning of case end here well enough.
So
```php
<?php
switch (time()):
    case 100: return false;
endswitch ;
```
gives end index 22, which is ';'

```php
<?php
switch (time()):
    case 100: return false;
endswitch ?> <?php echo 1;
```
gives end 20 which is `endswitch`

why aren't both `endswitch` here, or the first `;` and the second other `?>` ?

Comment (kubawerlos) [reply src/Tokenizer/Analyzer/SwitchAnalyzer.php:98]: The reasoning is: take the next statement and the previous token before it is the end of the current (the `switch`) one.

So for `switch ($x) { /* ... */ } nextStatement();` the end is `}`.
For `switch ($x): /* ... */ endswitch; nextStatement();` the end is `;` (after `endswitch`).
For `switch ($x): /* ... */ } endswitch ?>  <?php nextStatement();` the end is `endswitch` as `?>` is next statement - or is it not?

Comment (SpacePossum) [reply src/Tokenizer/Analyzer/SwitchAnalyzer.php:98]: I see, yeah, so from my POV I think for `switch ($x): /* ... */ } endswitch ?>` the close tag `?>` is what ends the `switch` , without it the statement is complete and invalid, I think `;` and `?>` have the same meaning here

Comment (kubawerlos) [reply src/Tokenizer/Analyzer/SwitchAnalyzer.php]: That makes sense - I've updated the tests and code.

Comment (SpacePossum) [reply src/Tokenizer/Analyzer/SwitchAnalyzer.php]: thanks!

**Comment by SpacePossum** [comment]
Thank you @kubawerlos.

**Pull request merged**